Add topic on MAVLink C library #defines (#492)
mavlink/mavlink-devguide: en/mavgen_c/defines.md

# MAVLink C Library #defines

The following C `#defines` can be set in code in order to tune the setup for use on different platforms.

> - `MAVLINK_USE_CONVENIENCE_FUNCTIONS` ([protocol.h](https://github.com/ArduPilot/pymavlink/blob/master/generator/C/include_v2.0/protocol.h)): Causes convenience functions to be defined, including: `_mav_finalize_message_chan_send()`, `_mavlink_send_uart()`, `_mavlink_resend_uart()`. To use, add `#define MAVLINK_USE_CONVENIENCE_FUNCTIONS` to your code.

Author: Why wouldn't you do this? Would there be any real saving in memory?

Reviewer: This just gives you a way of implementing these functions yourself. For instance in PX4 we need to send broadcast messages on all MAVLink instances, not just one, so the function is redefined to go through all instances and send off the data, see:
https://github.com/PX4/PX4-Autopilot/blob/6f861ba8897c67592627d4fe3011a4e19bef7942/src/modules/mavlink/mavlink_main.cpp#L88.

Author: Thanks @julianoes - I don't get that from the example - the linked example still takes a channel so sends on a particular instance rather than going through all of them. The equivalent convenience function does something similar https://github.com/ArduPilot/pymavlink/blob/15fc4778844d94ccf83d5f909cf45d07ac9dbfe4/generator/C/include_v2.0/mavlink_helpers.h#L1130

Could I get away with something like:

```suggestion
- `MAVLINK_USE_CONVENIENCE_FUNCTIONS` ([protocol.h](https://github.com/ArduPilot/pymavlink/blob/master/generator/C/include_v2.0/protocol.h)): Causes convenience functions to be defined, including: `_mav_finalize_message_chan_send()`, `_mavlink_send_uart()`, `_mavlink_resend_uart()`.
  Library users may choose not to set this and roll out their own implementations for efficiency (or other) reasons.
```

Reviewer: Ok.

> - `MAVLINK_COMM_NUM_BUFFERS`: Sets the maximum number of comms buffers to be used (comms channels).

Author: Would I be correct in thinking that you might specify `1` if you knew you had a mavlink component that would only have one connection to the network?

  By default this is set to 16 on Linux, Windows and macOS, and to 4 on other platforms.
  You might set `#define MAVLINK_COMM_NUM_BUFFERS 2` on an embedded system that would only ever have two channels, in order to reduce memory overheads.
  A stack overrun will occur if there are more buffers used than allocated.
  See [FAQ > How can I further reduce the generated C library size?](../about/faq.md#developers).
- `MAVLINK_SIGNING_FLAG_SIGN_OUTGOING` ([mavlink_types.h](https://github.com/ArduPilot/pymavlink/blob/master/generator/C/include_v2.0/mavlink_types.h)): Enable/disable outgoing signing.
  See [Enabling Signing on a Channel > Enabling Signing on a Channel](../mavgen_c/message_signing_c.md#enabling_signing_channel)

> - `MAVLINK_EXTERNAL_RX_STATUS` ([mavlink_helpers.h](https://github.com/ArduPilot/pymavlink/blob/master/generator/C/include_v2.0/mavlink_helpers.h)): `mavlink_status_t* mavlink_get_channel_status(uint8_t chan)` returns the status of a particular channel using a preallocated array.

Reviewer: ```suggestion
- `MAVLINK_EXTERNAL_RX_STATUS` ([mavlink_helpers.h](https://github.com/ArduPilot/pymavlink/blob/master/generator/C/include_v2.0/mavlink_helpers.h)): `mavlink_status_t* mavlink_get_channel_status(uint8_t chan)` returns the status of a particular channel using a preallocated `static` array.
```

  Define `MAVLINK_EXTERNAL_RX_STATUS` if you want to use your own external array (named `m_mavlink_status`) of `mavlink_status_t` objects.

> - `MAVLINK_EXTERNAL_RX_BUFFER` ([mavlink_helpers.h](https://github.com/ArduPilot/pymavlink/blob/master/generator/C/include_v2.0/mavlink_helpers.h)): Set if you want to define your own channel comms buffers.

Author: ```suggestion
- `MAVLINK_EXTERNAL_RX_BUFFER` ([mavlink_helpers.h](https://github.com/ArduPilot/pymavlink/blob/master/generator/C/include_v2.0/mavlink_helpers.h)): Set if you want to define your own channel comms buffers on the stack or heap, rather than using the pre-allocated `static` buffers.
```

>   If you define this you will need to declare an array of `mavlink_message_t` (one for each channel) named `m_mavlink_buffer`.

Author: Why would you provide your own status and buffers? They have to be the same type and hence size, so what would the benefit be?

Reviewer: If you don't want it allocated `static` but want to have it somewhere on the stack of heap instead.

Author: Great. I think I have made this clear.

- `NATIVE_BIG_ENDIAN` ([protocol.h](https://github.com/ArduPilot/pymavlink/blob/master/generator/C/include_v2.0/protocol.h)): Enable if using MAVLink on a system that is native big-endian (MAVLINK_LITTLE_ENDIAN is true by default).
- `MAVLINK_SEPARATE_HELPERS` ([protocol.h](https://github.com/ArduPilot/pymavlink/blob/master/generator/C/include_v2.0/protocol.h)): Set to remove all the helpers declared `mavlink_helpers.h`, allowing you to provide an alternative implementation.
- `MAVLINK_MAX_PAYLOAD_LEN`: Override the maximum payload length.

>   The default maximum payload length is 255 bytes. With care you can override this to reduce memory usage.

Author: ```suggestion
  The default maximum payload length is 255 bytes. With care you can override this to reduce the size, and hence memory usage.
```

>   For example, to reduce the maximum payload to the smallest size supported by a dialect you could do:

Reviewer: Maybe worth noting that you can't make it longer because the size field is only a `uint8_t`.

Author: Done
